### PR TITLE
#97: Added soft abort to installation if knative exists

### LINKED ISSUE #97: Delete old knative resources when installing on an existing cluster
If the quickstart plugin detects an existing Kind cluster, it prompt the user with the choice deleting and recreating the cluster. If the user chooses to NOT delete their existing cluster, there can be conflicts when installing Knative over top of an existing installation. 

It might be a good idea to check to see if the knative namespaces exist, and if so to re-prompt the user to either delete / recreate the cluster or abort.

### PR DESCRIPTION
**Pull request by ehudyonasi**
# Changes

- :gift: Added soft abortion to kn-quickstart if the knative namespaces already exists on clusters.
-
-
-

/kind <kind>

Fixes #97 

**Release Note**

```release-note

```

**Docs**

```docs

```

### CODE REVIEW

**Comment by ehudyonasi** [comment]
@pkosiec  let me know what you think of it

File: pkg/minikube/minikube.go [hunk 1]
```diff
@@ -154,6 +155,22 @@ func checkForExistingCluster() error {
...
 			fmt.Println("Installation skipped")
+			checkKnativeNamespace := exec.Command("kubectl", "get", "namespaces")
+			output, err := checkKnativeNamespace.CombinedOutput()
+			namespaces := string(output)
+			if err != nil {
+				fmt.Println(string(output))
+				return fmt.Errorf("check existing cluster: %w", err)
+			}
+			if strings.Contains(namespaces, "knative") {
+				fmt.Print("Knative installation already exists.\nIt is advised to delete and recreate the cluster [y/N]: ")
```
Comment (psschwei) [inline pkg/minikube/minikube.go:166]: Let's tweak the wording a bit

```suggestion
				fmt.Print("Knative installation already exists.\nDelete and recreate the cluster [y/N]: ")
```

File: pkg/minikube/minikube.go [hunk 2]
```diff
@@ -154,6 +155,22 @@ func checkForExistingCluster() error {
...
+				fmt.Scanf("%s", &resp)
+				if strings.ToLower(resp) != "y" {
+					fmt.Println("Skipping installation")
+					installKnative = false
+					return nil
+				}
```
Comment (psschwei) [inline pkg/minikube/minikube.go:172]: if they do respond `y`, then we need to run the delete/recreate cluster bits

```suggestion
				if strings.ToLower(resp) != "y" {
					fmt.Println("Skipping installation")
					installKnative = false
					return nil
				} else {
				/* do stuff */
				}
```

File: pkg/kind/kind.go [hunk 3]
```diff
@@ -203,23 +204,31 @@ func checkForExistingCluster() error {
...
 			fmt.Println("\n    Installation skipped")
+			checkKnativeNamespace := exec.Command("kubectl", "get", "namespaces")
+			output, err := checkKnativeNamespace.CombinedOutput()
+			namespaces := string(output)
+			if err != nil {
+				fmt.Println(string(output))
+				return fmt.Errorf("check existing cluster: %w", err)
+			}
+			if strings.Contains(namespaces, "knative") {
+				fmt.Print("Knative installation already exists.\nIt is advised to delete and recreate the cluster [y/N]: ")
```
Comment (psschwei) [inline pkg/kind/kind.go:220]:
```suggestion
				fmt.Print("Knative installation already exists.\nDelete and recreate the cluster [y/N]: ")
```

File: pkg/kind/kind.go [hunk 4]
```diff
@@ -234,6 +243,26 @@ func checkForExistingCluster() error {
...
+	}
+	deleteContainerRegistry := deleteContainerRegistry()
+	if err := deleteContainerRegistry.Run(); err != nil {
+		return fmt.Errorf("delete container registry: %w", err)
+	}
+	if err := createNewCluster(); err != nil {
+		return fmt.Errorf("new cluster: %w", err)
+	}
+	if err := connectLocalRegistry(); err != nil {
+		return fmt.Errorf("local-registry: %w", err)
```
Comment (psschwei) [inline pkg/kind/kind.go:261]: When trying to recreate the cluster, I got an error here with a `failed to connect local registry to kind cluster` error

File: pkg/minikube/minikube.go (hunk 1 above)
Comment (ehudyonasi) [reply pkg/minikube/minikube.go:166]: Sure, sounds good.

File: pkg/minikube/minikube.go (hunk 2 above)
Comment (ehudyonasi) [reply pkg/minikube/minikube.go:172]: Right, will add it

File: pkg/kind/kind.go (hunk 4 above)
Comment (ehudyonasi) [reply pkg/kind/kind.go:261]: Will look into it

Comment (ehudyonasi) [reply pkg/kind/kind.go:264]: Ok, I've found the issue. comitted the changes

File: pkg/minikube/minikube.go [hunk 5]
```diff
@@ -154,6 +155,32 @@ func checkForExistingCluster() error {
...
+			if strings.Contains(namespaces, "knative") {
+				fmt.Print("Knative installation already exists.\nDelete and recreate the cluster [y/N]: ")
+				fmt.Scanf("%s", &resp)
+				if strings.ToLower(resp) != "y" {
+					fmt.Println("Skipping installation")
+					installKnative = false
+					return nil
+				} else {
+					fmt.Println("deleting cluster...")
+					deleteCluster := exec.Command("minikube", "delete", "--profile", "--insecure-registry", clusterName)
```
Comment (psschwei) [inline pkg/minikube/minikube.go:174]:
```suggestion
					deleteCluster := exec.Command("minikube", "delete", "--profile", clusterName)
```

don't need to specify the registry on delete (sorry if I missed this on an earlier review...)

File: pkg/minikube/minikube.go [hunk 6]
```diff
@@ -154,6 +155,32 @@ func checkForExistingCluster() error {
...
+					if err := deleteCluster.Run(); err != nil {
+						return fmt.Errorf("delete cluster: %w", err)
+					}
+					if err := createNewCluster(); err != nil {
+						return fmt.Errorf("new cluster: %w", err)
+					}
```
Comment (psschwei) [inline pkg/minikube/minikube.go:180]: let's pull this into a separate function to use here and lines 187-93

Comment (psschwei) [reply pkg/minikube/minikube.go]: > let's pull this into a separate function to use here and lines 187-93

still need this one

Comment (ehudyonasi) [reply pkg/minikube/minikube.go]: Right, will add it today.

**Review by psschwei — COMMENTED** [review]
Looks good... thanks for the PR!

/lgtm
/approve

**Pull request merged**